### PR TITLE
Add password reset

### PR DESCRIPTION
**Pull request by antoalconti**
- Reset controller was added.

- Reset create and update Spects was added

### CODE REVIEW

File: config/locales/devise.en.yml [hunk 1]
```diff
@@ -32,10 +32,16 @@ en:
...
       no_token: "You can't access this page without coming from a password reset email. If you do come from a password reset email, please make sure you used the full URL provided."
+      nonexisten_email: "Unable to find user with email '%{email}'."
       send_instructions: "You will receive an email with instructions on how to reset your password in a few minutes."
+      send_reset_instructions: "An email has been sent to '%{email}' containing instructions for resetting your password."
       send_paranoid_instructions: "If your email address exists in our database, you will receive a password recovery link at your email address in a few minutes."
       updated: "Your password has been changed successfully. You are now signed in."
+      successfully_updated: "Your password has been successfully updated."
       updated_not_active: "Your password has been changed successfully."
+      empty_parameter: "You must fill out the fields labeled 'Password' and 'Password confirmation'."
+      passwords_do_not_match: "Password confirmation doesn't match Password"
```
Comment (dani24gl) [inline config/locales/devise.en.yml]: Maybe `doesn't match with Password`

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    it 'updates the user reset password token' do
+      expect(user.reload.reset_password_token).to be_present
+      subject
+      expect(user.reload.reset_password_token).to be_nil
+    end
+  end
+
+  context 'when the request is not valid' do
+    context 'when the user does not request rest password' do
```
Comment (dani24gl) [inline spec/requests/api/v1/passwords/update_spec.rb]: nit: typo missing `e` in reset

File: app/controllers/api/v1/passwords_controller.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module Api
+  module V1
+    class PasswordsController < DeviseTokenAuth::PasswordsController
+      include Api::Concerns::ActAsApiRequest
+      skip_before_action :check_json_request, on: :edit
```
Comment (sebaherrera07) [inline app/controllers/api/v1/passwords_controller.rb]: Why was this necessary?

File: db/migrate/20191203155817_add_allow_password_to_user.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddAllowPasswordToUser < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :allow_password_change, :boolean, default: false, null: false
```
Comment (sebaherrera07) [inline db/migrate/20191203155817_add_allow_password_to_user.rb]: Why was this necessary?

File: config/locales/devise.en.yml [hunk 5]
```diff
@@ -32,10 +32,16 @@ en:
       success: "Successfully authenticated from %{kind} account."
     passwords:
       no_token: "You can't access this page without coming from a password reset email. If you do come from a password reset email, please make sure you used the full URL provided."
+      nonexisten_email: "Unable to find user with email '%{email}'."
```
Comment (sebaherrera07) [inline config/locales/devise.en.yml]: Are this custom messages? In that case I think it would be better to add them on `en.yml` file instead of `devise.en.yml` so that we leave the one of devise just with the default messages.

Comment (sebaherrera07) [inline config/locales/devise.en.yml]: `non_existent_email`  -> You missed a `t` . 

Also, usually web applications shouldn't say when an email exists or not. Depending on what the user is trying to do, you may show a more generic message.

File: config/locales/devise.en.yml [hunk 6]
```diff
@@ -32,10 +32,16 @@ en:
...
       send_instructions: "You will receive an email with instructions on how to reset your password in a few minutes."
+      send_reset_instructions: "An email has been sent to '%{email}' containing instructions for resetting your password."
       send_paranoid_instructions: "If your email address exists in our database, you will receive a password recovery link at your email address in a few minutes."
       updated: "Your password has been changed successfully. You are now signed in."
+      successfully_updated: "Your password has been successfully updated."
       updated_not_active: "Your password has been changed successfully."
+      empty_parameter: "You must fill out the fields labeled 'Password' and 'Password confirmation'."
+      doesn't_match_with_Password: "Password confirmation doesn't match Password"
```
Comment (sebaherrera07) [inline config/locales/devise.en.yml]: Usually we don't put `'` in key values, and you also put a capital `P`.

I would suggest calling this key: `doesnt_match_with_password`

File: config/locales/devise.en.yml [hunk 7]
```diff
@@ -32,10 +32,16 @@ en:
...
+      password_length: "Password is too short (minimum is 6 characters)"
```
Comment (sebaherrera07) [inline config/locales/devise.en.yml]: The number 6 here is hardocded. Usually we would pass a parameter to the translation, so that number can dinamically change.

File: spec/requests/api/v1/passwords/create_spec.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      subject
+      message = I18n.t('devise.passwords.nonexisten_email', params { :email })
+      expect(json).to include_json(errors: [message])
+    end
+
+    it 'returns a not found status' do
+      subject
+      expect(response).to have_http_status(:not_found)
+    end
+  end
```
Comment (sebaherrera07) [inline spec/requests/api/v1/passwords/create_spec.rb]: Maybe you can add a test to check that the `user.reload.reset_password_token` is `nil` after calling `subject`

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+RSpec.describe 'PATCH /api/v1/users/password', type: :request do
+  let(:user) { create(:user) }
+  let(:password_token) { user.send(:set_reset_password_token) }
+  let!(:headers) do
+    params = {
+      reset_password_token: password_token,
+      redirect_url: ENV['PASSWORD_RESET_URL']
+    }
+    get edit_user_password_path, params: params, headers: auth_headers
```
Comment (sebaherrera07) [inline spec/requests/api/v1/passwords/update_spec.rb]: Maybe this `get` and the `edit_response_params` shouldn't be inside this `let`. 

Can you try moving them to a `before` action? or maybe inside each test itself.

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      'client' => edit_response_params['client_id']
+    }
+  end
+
+  subject { patch user_password_url, params: params, headers: headers, as: :json }
+
+  context 'when the request is valid' do
+    let(:params) do
+      {
+        "password": 'Newpass123',
```
Comment (sebaherrera07) [inline spec/requests/api/v1/passwords/update_spec.rb]: I think you can just put `password: 'Newpass123'` (you can avoid putting the double quotes for key values)

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+      it 'returns a unauthorized status' do
+        subject
+        expect(response).to have_http_status(:unauthorized)
+      end
+    end
+
+    context 'when params is missing' do
+      let(:params) do
+        {
+          "password_confirmation": 'Newpass123'
```
Comment (sebaherrera07) [inline spec/requests/api/v1/passwords/update_spec.rb]: I think you can just put `password_confirmation: 'Newpass123'` (you can avoid putting the double quotes for key values)

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      it 'returns a unprocessable entity status' do
+        subject
+        expect(response).to have_http_status(:unprocessable_entity)
+      end
+    end
+
+    context 'when the password_confirmation does not match password' do
+      let(:params) do
+        {
+          "password": 'Newpass1',
```
Comment (sebaherrera07) [inline spec/requests/api/v1/passwords/update_spec.rb]: You can avoid putting the double quotes for key values

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+      it 'returns a unprocessable entity status' do
+        subject
+        expect(response).to have_http_status(:unprocessable_entity)
+      end
+    end
+
+    context 'when the password length is too short' do
+      let(:params) do
+        {
+          "password": 'Newp',
```
Comment (sebaherrera07) [inline spec/requests/api/v1/passwords/update_spec.rb]: You can avoid putting the double quotes for key values

File: db/migrate/20191203155817_add_allow_password_to_user.rb (hunk 4 above)
Comment (antoalconti) [reply db/migrate/20191203155817_add_allow_password_to_user.rb]: Because devise need it, devise set the value on true when the user ask for a reset token and check the boolean value before update the password.

File: app/controllers/api/v1/passwords_controller.rb (hunk 3 above)
Comment (antoalconti) [reply app/controllers/api/v1/passwords_controller.rb]: I understand that edit request in a link, does not include a json on the request.

File: config/locales/devise.en.yml (hunk 7 above)
Comment (antoalconti) [reply config/locales/devise.en.yml]: Thank you, will be helpful in future. This one is a devise message.

File: spec/requests/api/v1/passwords/create_spec.rb (hunk 8 above)
Comment (antoalconti) [reply spec/requests/api/v1/passwords/create_spec.rb]: In this context there is not a user, i did not need it for this tests; for that i did not check it.
